# CellTagWorkflow

This repository will contain our CellTag workflow.

Here is the link to the GEO DataSet which contains all of the sequencing data we generated.

https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99915

From these data we will select one timepoint to use as an example of our CellTag processing.

We will use data collected from Timecourse 1 at Day 15. The link to the SRA for this sample is [here](https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR7347033).

# CellTag Extraction and Quantification

First we will download the BAM file generated by the 10x CellRanger pipeline for this sample.

The link for this BAM file is:

https://sra-download.ncbi.nlm.nih.gov/traces/sra65/SRZ/007347/SRR7347033/hf1.d15.possorted_genome_bam.bam


```bash
# Not Run
wget https://sra-download.ncbi.nlm.nih.gov/traces/sra65/SRZ/007347/SRR7347033/hf1.d15.possorted_genome_bam.bam

```

Now we can extract the CellTag reads from the BAM file we just downloaded. We do this using samtools and grep.


```bash

samtools view hf1.d15.possorted_genome_bam.bam | grep -P 'GGT[ACTG]{8}GAATTC' > v1.celltag.reads.out

```

With the CellTag reads extracted we use a custom gawk script to parse the file and retain only the information we need.


```bash

./scripts/celltag.parse.reads.10x.sh -v tagregex="CCGGT([ACTG]{8})GAATTC" v1.celltag.reads.out > v1.celltag.parsed.tsv

```

Now with the parsed reads we will create a Cell by CellTag Matrix.


```bash

Rscript ./scripts/matrix.count.celltags.R ./cell.barcodes/hf1.d15.barcodes.tsv v1.celltag.parsed.tsv hf1.d15.v1

```


# Clone Calling

Load functions required for clone calling


```r
library(igraph)
library(proxy)
library(corrplot)
library(data.table)

source("./scripts/CellTagCloneCalling_Function.R")
```


Load CellTag Matrix and create binary matrix.



```r
celltag.mat <- readRDS("./hf1.d15.v1.celltag.matrix.Rds")

rownames(celltag.mat) <- celltag.mat$Cell.BC

celltag.mat <- celltag.mat[, -1]

celltag.bin <- SingleCellDataBinarization(celltag.dat = celltag.mat, 2)
```


Filter CellTag matrix


```r
celltag.filt <- SingleCellDataWhitelist(celltag.dat = celltag.bin, whitels.cell.tag.file = "whitelist/V1.CellTag.Whitelist.csv")

celltag.filt <- MetricBasedFiltering(whitelisted.celltag.data = celltag.filt, cutoff = 20, comparison = "less")

celltag.filt <- MetricBasedFiltering(whitelisted.celltag.data = celltag.filt, cutoff = 2, comparison = "greater")
```


Now Call Clone based on Jaccard Similarity


```r
celltag.jac <- JaccardAnalysis(whitelisted.celltag.data = celltag.filt, plot.corr = FALSE)

clones <- CloneCalling(Jaccard.Matrix = celltag.jac, output.dir = "./", output.filename = "hf1.d15.v1.clones.csv", correlation.cutoff = 0.7)
```


# Lineage and Network Visualization

