# CellTagWorkflow

This repository will contain our CellTag workflow.

Here is the link to the GEO DataSet which contains all of the sequencing data we generated.

https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99915

From these data we will select one timepoint to use as an example of our CellTag processing.

We will use data collected from Timecourse 1 at Day 15. The link to the SRA for this sample is [here](https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR7347033).

First we will download the BAM file generated by the 10x CellRanger pipeline for this sample.

The link for this BAM file is:

https://sra-download.ncbi.nlm.nih.gov/traces/sra65/SRZ/007347/SRR7347033/hf1.d15.possorted_genome_bam.bam


```bash
# Not Run
wget https://sra-download.ncbi.nlm.nih.gov/traces/sra65/SRZ/007347/SRR7347033/hf1.d15.possorted_genome_bam.bam

```

Now we can extract the CellTag reads from the BAM file we just downloaded. We do this using samtools and grep.


```bash

samtools view hf1.d15.possorted_genome_bam.bam | grep -p GATTACA[ACTG]{8}GATTACA > celltag.reads.out

```

With the CellTag reads extracted we use a custom gawk script to parse the file and retain only the information we need.


```bash

../scripts/parse.tag.reads -v 

```


# CellTag Extraction and Quantification

# Clone Calling

# Lineage and Network Visualization

